fdbserver/RestoreApplier.actor.cpp

```diff
@@ -134,102 +134,253 @@ ACTOR static Future<Void> handleSendMutationVectorRequest(RestoreSendMutationVec
 		try {
-			tr->reset();
-			tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
-			tr->setOption(FDBTransactionOptions::LOCK_AWARE);
-			transactionSize = 0;
-
-			for (; it != self->kvOps.end(); ++it) {
-				numVersion++;
-				//TraceEvent("FastRestore").detail("Applier", self->id()).detail("ApplyKVsToDBVersion", it->first);
-				state MutationRef m;
-				for (; index < it->second.size(); ++index) {
-					m = it->second[index];
+			// Check if the transaction succeeds
+			if (progress.lastTxnHasError) {
+				tr->reset();
+				tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+				tr->setOption(FDBTransactionOptions::LOCK_AWARE);
+				Optional<Value> txnSucceeded = wait(tr->get(restoreApplierKeyFor(self->id(), progress.curTxnId)));
+				if (!txnSucceeded.present()) {
+					TraceEvent(SevWarn, "FastRestore_ApplyTxnError")
+					    .detail("TxnStatusFailed", progress.curTxnId)
+					    .detail("ApplierApplyToDB", self->id())
+					    .detail("UncommittedTxnId", progress.uncommittedTxnId)
+					    .detail("CurIteratorVersion", progress.curItInCurTxn->first)
+					    .detail("StartIteratorVersionInUncommittedTxn", progress.startItInUncommittedTxn->first)
+					    .detail("CurrentIndexInFailedTxn", progress.curIndexInCurTxn)
+					    .detail("StartIndexInUncommittedTxn", progress.startIndexInUncommittedTxn)
+					    .detail("NumIncludedAtomicOps", progress.numAtomicOps);
+					progress.rollback();
+					continue;
+				} else {
+					TraceEvent(SevWarn, "FastRestore_ApplyTxnError")
+					    .detail("TxnStatusSucceeded", progress.curTxnId)
+					    .detail("ApplierApplyToDB", self->id())
+					    .detail("CurIteratorVersion", progress.curItInCurTxn->first)
+					    .detail("CurrentIteratorMutations", progress.curItInCurTxn->second.size())
+					    .detail("CurrentIndexInSucceedTxn", progress.curIndexInCurTxn)
+					    .detail("NumIncludedAtomicOps", progress.numAtomicOps);
+
+					// Txn succeeded and exectue the same logic when txn succeeds
+				}
+			} else { // !lastTxnHasError: accumulate mutations in a txn
+				tr->reset();
+				tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+				tr->setOption(FDBTransactionOptions::LOCK_AWARE);
+				TraceEvent("FastRestore_ApplierTxn")
+				    .detail("ApplierApplyToDB", self->id())
+				    .detail("TxnId", progress.curTxnId)
+				    .detail("StartIndexInCurrentTxn", progress.curIndexInCurTxn)
+				    .detail("CurrentIteratorMutations", progress.curItInCurTxn->second.size())
+				    .detail("Version", progress.curItInCurTxn->first);
+
+				// restoreApplierKeyFor(self->id(), curTxnId) to tell if txn succeeds at an unknown error
+				tr->set(restoreApplierKeyFor(self->id(), progress.curTxnId), restoreApplierTxnValue);
+
+				while (1) { // Loop: Accumulate mutations in a transaction
+					state MutationRef m;
+					ASSERT_WE_THINK(progress.curIndexInCurTxn < progress.curItInCurTxn->second.size());
+
+					m = progress.curItInCurTxn->second[progress.curIndexInCurTxn];
```

Comment:
Since you have an abstract `progress`, it's better to use an API, like `getCurrentMutation()`.